snagrecover: imx: add support for i.MX53

mfuzzey opened the pull request:
i.MX53 devices use a different version of the Serial Download Protocol.

Raw USB bulk endpoints are used rather than HID (so no report IDs)

The following commands are not implemented:
	DCD_WRITE
	JUMP_ADDRESS
	SKIP_DCD_HEADER

Replace DCD_WRITE by manually interpreting the DCD data as individual read/write commands (as imx_usb_loader does).

Tested on a custom i.MX53 based board and regression tested on a Digi Connect Core 6 i.MX6D evaluation board.

rgantois commented on src/snagrecover/recoveries/imx.py:
```diff
@@ -77,17 +77,70 @@
+	if dev is None:
+		access_error("USB RAW", f"{vid:04x}:{pid:04x}")
+
+	class Adapter:
```
Could you please move this class definition outside of the function body?

rgantois commented on src/snagrecover/protocols/imx_sdp.py:
```diff
@@ -145,49 +147,157 @@ def write32(self, addr: int, value: int) -> bool:
+		while pos < end:
+			tag, lg, param = struct.unpack(">BHB", blob[pos:pos + 4])
+
+			def invoke_for_each_addr_data(fn, param=param, pos=pos, lg=lg):
```
I find this a bit hard to read. Could you please move this function definition outside of the SDPCommand class def? I think it would be more appropriate since it doesn't explicitly use the class attributes. Plus, it would tame the indentation a bit.

rgantois reviewed (changes requested):
Wow thanks a lot, that is a really nice contribution! I just have a couple of nitpicks but apart from those the code is OK for me.

mfuzzey commented:
Hi,
Thanks for the quick review.

I could do the changes you request but I'm not convinced they would really improve the code.

For the Adapter class defined in a function it's encapsulation. It reduces the mental overload when reading the code since you immeadiately know that class is only used locally. If it were a top level class you'd have to ask where else it is used.

For invoke_for_each_addr_data() it can't just be moved outside as is since it also requires access to nested scope (in particular blob). That would have to be added to the function parameter list and the param, pos and lg parameters would have to be explicitly passed at the two function invocation call sites (so increasing the number of parameters passed at the call site from 1 to 5). Currently they are explicitly bound at the function definition which isn't even strictly necessary as they are in scope in the loop. But I did that to placate a false positive from the linter.

This type of style is really fairly idiomatic python even though it can look strange for those more used to other languages where it's not possible.
See https://realpython.com/inner-functions-what-are-they-good-for/

And "import this" of course gives.
"Namespaces are one honking great idea -- let's do more of those!"

All that being said at the end of the day you are the maintainer of this code so if you're really not comfortable with this style I'll change it.

I'm on holiday this evening though so it won't be until September in that case (just so if you don't hear from me in the next few weeks you know I'm not annoyed just not there).
So have a think and if you still prefer I change it I'll do it in September.

rgantois commented:
Thanks for the precisions, the `Adapter` class can stay in the def if you think it is better that way, I don't really mind.
However, I must insist on `invoke_for_each_addr_data` being moved out of the def. To me, this solution is a bit too complex and difficult to parse, so I'd rather go with a simpler albeit less elegant solution. I personally don't mind passing 5 parameters if we have to.
And of course, take all the time you need, I'm planning on releasing 1.2 soon so it suits me better if we postpone substantial code changes to September.

mfuzzey commented:
New version pushed that extracts invoke_for_each_addr_data() [now as _invoke_for_each_dcd_element_addr_data()].
I've also added a helper _unpack_from_blob() to avoid duplication in stuff like struct.unpack(">BHB, blob[pos:pos+4]) where the +4 can be inferred from the format string.
And rebased to master.

rgantois commented:
Great, thanks a lot!
